compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -513,18 +513,26 @@ impl<'f> PerFunctionContext<'f> {
             Instruction::ArrayGet { array, .. } => {
                 let result = self.inserter.function.dfg.instruction_results(instruction)[0];
 
-                let array = *array;
-                let array_typ = self.inserter.function.dfg.type_of_value(array);
-                if array_typ.contains_reference() {
+                if self.inserter.function.dfg.type_of_value(result).contains_reference() {
+                    let array = *array;
                     self.instruction_input_references
                         .extend(references.get_aliases_for_value(array).iter());
                     references.mark_value_used(array, self.inserter.function);
 
-                    let expression = Expression::ArrayElement(array);
+                    // An expression for the array might already exist, so try to fetch it first
+                    let expression = references.expressions.get(&array).copied();
+                    let expression = expression.unwrap_or(Expression::ArrayElement(array));
```

Comment:
I think this is not correct - ArrayElement would imply we're getting the element of an array e.g. if we were looking up `result` instead. We're getting the array itself here though so it should probably just be Other.

Reply:
Ah, no, it's still used in MakeArray...